File: sdks/python/relayhq/subagent.py
"""subagent() — turn an Agent into a tool another Agent can call.

When the parent's LLM invokes the subagent, this handler runs
`child.run(input)` linked to the parent run via parent_run_id and
workflow_id. The whole tree shares one workflow_id so the dashboard can
render the call graph and aggregate cost.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from .agent import Agent
from .tool import tool
from .types import FunctionTool


MAX_SUBAGENT_DEPTH = 5

_depth_by_workflow: Dict[str, int] = {}


def subagent(
    *,
    name: str,
    description: str,
    agent: Agent,
    input_schema: Optional[Dict[str, Any]] = None,
    max_depth: int = MAX_SUBAGENT_DEPTH,
) -> FunctionTool:
    """Wrap an Agent as a callable tool.

    Default input shape is `{"input": str}`; override `input_schema` if
    your sub-agent needs a structured prompt.

    Example:

        researcher = create_agent(model="gpt-4o", tools=[web_search])
        writer = create_agent(
            model="claude-sonnet-4-6",
            tools=[
                subagent(
                    name="research",
                    description="Research a topic",
                    agent=researcher,
                ),
            ],
        )
    """
    schema = input_schema or {
        "type": "object",
        "properties": {
            "input": {"type": "string", "description": "Prompt for the sub-agent"}
        },
        "required": ["input"],
        "additionalProperties": False,
    }

    async def handler(input: Any, ctx: Optional[Dict[str, Any]] = None) -> Any:
        prompt = _extract_prompt(input)
        if prompt is None:
            return {"error": "subagent expected { input: string }"}

        workflow_key = (ctx or {}).get("workflow_id") or (ctx or {}).get("run_id") or "anon"
        current = _depth_by_workflow.get(workflow_key, 0)
        if current >= max_depth:
            return {
                "error": f"subagent depth limit reached ({max_depth}); refusing to recurse"
            }
        _depth_by_workflow[workflow_key] = current + 1

        try:
            parts: list[str] = []
            usage = None
            last_error = None
            async for ev in agent.run(
                prompt,
                parent_run_id=(ctx or {}).get("run_id"),
                workflow_id=(ctx or {}).get("workflow_id") or (ctx or {}).get("run_id"),
            ):
                if ev.get("type") == "token":
                    parts.append(ev.get("text", ""))
                elif ev.get("type") == "done":
                    parts = [ev.get("output", "")]
                    usage = ev.get("usage")
                elif ev.get("type") == "error":
                    last_error = ev.get("message")
            if last_error and not parts:
                return {"error": last_error}
            return {"output": "".join(parts), "usage": usage}
        finally:
            after = _depth_by_workflow.get(workflow_key, 1) - 1
            if after <= 0:
                _depth_by_workflow.pop(workflow_key, None)
            else:
                _depth_by_workflow[workflow_key] = after

    return tool(
        name=name,
        description=description,
        input_schema=schema,
        handler=handler,
    )


def _extract_prompt(input: Any) -> Optional[str]:
    if isinstance(input, str):
        return input
    if isinstance(input, dict) and isinstance(input.get("input"), str):
        return input["input"]
    return None

File: packages/sdk/src/subagent.ts
import type { Agent } from "./agent.js";
import { tool } from "./tool.js";
import type { AgentEvent, FunctionTool, ToolContext } from "./types.js";

// ─── subagent() ─────────────────────────────────────────────────────────────
//
// Turn an Agent into a Tool that another Agent can call. When the parent's
// LLM decides to invoke the sub-agent, this handler runs `child.run(input)`
// linked to the parent run, drains the events, and returns the final output
// as the tool result. The whole tree of runs shares one workflow_id, so the
// dashboard can render the call graph and aggregate cost.
//
// Usage:
//   const researcher = createAgent({ model: "gpt-4o", tools: [web_search] });
//   const writer = createAgent({
//     model: "claude-sonnet-4-6",
//     tools: [
//       subagent({
//         name: "research",
//         description: "Research a topic and return key findings",
//         agent: researcher,
//       }),
//     ],
//   });
//
// Notes:
//   - `input` to the tool is `{ input: string }` by default; override the
//     schema if you want different fields. The handler reads `input` from
//     the args.
//   - Events from the sub-agent are NOT yielded into the parent stream —
//     only the final output is returned as the tool result. If you want to
//     surface progress, the parent can subscribe to the dashboard's
//     workflow tree endpoint (`GET /v1/workflows/:id`).
//   - Depth is capped at MAX_SUBAGENT_DEPTH to prevent runaway recursion.
//     Each subagent invocation increments depth via the SUBAGENT_DEPTH_KEY
//     in ctx (carried across via a closure on the handler).

const MAX_SUBAGENT_DEPTH = 5;

export type SubagentSpec = {
  name: string;
  description: string;
  agent: Agent;
  // Override the input schema. Default expects { input: string }.
  inputSchema?: Record<string, unknown>;
  // Cap recursion depth for THIS subagent specifically. Defaults to the
  // global ceiling.
  maxDepth?: number;
};

// Internal: thread-local-ish counter via WeakMap keyed on workflowId.
// Single-threaded JS means a simple Map keyed by workflow is fine.
const depthByWorkflow = new Map<string, number>();

export function subagent(spec: SubagentSpec): FunctionTool {
  const inputSchema = spec.inputSchema ?? {
    type: "object",
    properties: {
      input: { type: "string", description: "Prompt for the sub-agent" },
    },
    required: ["input"],
    additionalProperties: false,
  };
  const maxDepth = spec.maxDepth ?? MAX_SUBAGENT_DEPTH;

  return tool({
    name: spec.name,
    description: spec.description,
    inputSchema,
    handler: async (input: unknown, ctx?: ToolContext) => {
      const prompt = extractPrompt(input);
      if (!prompt) {
        return { error: "subagent expected { input: string }" };
      }

      const workflowKey = ctx?.workflowId ?? ctx?.runId ?? "anon";
      const current = depthByWorkflow.get(workflowKey) ?? 0;
      if (current >= maxDepth) {
        return {
          error: `subagent depth limit reached (${maxDepth}); refusing to recurse`,
        };
      }
      depthByWorkflow.set(workflowKey, current + 1);

      try {
        const sub = spec.agent.run(prompt, {
          parentRunId: ctx?.runId,
          workflowId: ctx?.workflowId ?? ctx?.runId,
          signal: ctx?.signal,
        });

        const segments: string[] = [];
        let usage: { input_tokens: number; output_tokens: number } | null = null;
        let lastError: string | null = null;

        for await (const ev of sub as AsyncIterable<AgentEvent>) {
          if (ev.type === "token") {
            segments.push(ev.text);
          } else if (ev.type === "done") {
            // The full output is also in ev.output; prefer it over the
            // streamed token sum since it reflects the model's final form.
            segments.length = 0;
            segments.push(ev.output);
            usage = ev.usage ?? null;
          } else if (ev.type === "error") {
            lastError = ev.message;
          }
        }

        if (lastError && segments.length === 0) {
          return { error: lastError };
        }

        return {
          output: segments.join(""),
          usage,
        };
      } finally {
        const after = (depthByWorkflow.get(workflowKey) ?? 1) - 1;
        if (after <= 0) {
          depthByWorkflow.delete(workflowKey);
        } else {
          depthByWorkflow.set(workflowKey, after);
        }
      }
    },
  });
}

function extractPrompt(input: unknown): string | null {
  if (typeof input === "string") return input;
  if (input && typeof input === "object" && "input" in input) {
    const v = (input as { input?: unknown }).input;
    if (typeof v === "string") return v;
  }
  return null;
}
